File: leetcode/python/705_design_hash_set.py
class MyHashSet:
    def __init__(self):
        self.arr = [[] for _ in range(8)]
        self.len: int = 0

    def add(self, key: int) -> None:
        if self.contains(key):
            return
        self.len += 1
        if self.len >= len(self.arr) // 2:
            self._migrate()
        i = self._hash(key)
        self.arr[i].append(key)

    def remove(self, key: int) -> None:
        if not self.contains(key):
            return
        i = self._hash(key)
        self.arr[i].remove(key)
        self.len -= 1

    def contains(self, key: int) -> bool:
        i = self._hash(key)
        for n in self.arr[i]:
            if n == key:
                return True
        return False

    def _hash(self, k: int) -> int:
        return k % len(self.arr)

    def _migrate(self) -> None:
        arr = [[] for _ in range(len(self.arr) * 2)]
        for bucket in self.arr:
            for n in bucket:
                i = n % len(arr)
                arr[i].append(n)
        self.arr = arr

h = MyHashSet()
h.add(1)      
h.add(2)      
print(h.contains(1))
print(h.contains(3))
h.add(2)
print(h.contains(2))
h.remove(2)
print(h.contains(2))

File: leetcode/rust/src/705_design_hash_set.rs
struct MyHashSet {
    // buckets
    buckets: Vec<Vec<i32>>,
    size: usize,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyHashSet {
    fn new() -> Self {
        let size = 1000;
        Self {
            size,
            buckets: vec![vec![]; size],
        }
    }

    fn hash(&self, key: i32) -> usize {
        key as usize % self.size
    }

    fn expand(&mut self) {
        let old_size = self.size;
        self.size *= 2;
        let mut new_buckets = vec![vec![]; self.size];
        for i in 0..old_size {
            for key in &self.buckets[i] {
                let hash = self.hash(*key);
                if !new_buckets[hash].contains(key) {
                    new_buckets[hash].push(*key);
                }
            }
        }
        self.buckets = new_buckets;
    }
    
    pub fn add(&mut self, key: i32) {
        let hash = self.hash(key);
        if !self.buckets[hash].contains(&key) {
            self.buckets[hash].push(key);
        }

        // increase size if needed
        if self.buckets[hash].len() > self.size / 2 {
            self.expand();
        }
    }
    
    pub fn remove(&self, key: i32) {
        let hash = self.hash(key);
        if let Some(index) = self.buckets[hash].iter().position(|&x| x == key) {
            self.buckets[hash].remove(index);
        }
    }
    
    pub fn contains(&self, key: i32) -> bool {
        let hash = self.hash(key);
        self.buckets[hash].contains(&key)
    }
}
